### ISSUE TITLE
Azure authorization exception using Flask Dance

### ISSUE THREAD

**Issue by leonbrag** [singingwolfboy/flask-dance#149]
I have following code running in Docker container in Azure:

```
from flask import Flask, redirect, url_for
from flask_dance.contrib.azure import make_azure_blueprint, azure

import time
app = Flask(__name__)



app.secret_key = "supersekrit"
blueprint = make_azure_blueprint(
    client_id="foo",
    client_secret="bar",
)
app.register_blueprint(blueprint, url_prefix="/login")

@app.route('/')
def hello_world():
	if not azure.authorized:
		return redirect(url_for("azure.login"))
	resp = azure.get("/v1.0/me")
	assert resp.ok
	return resp.json()["mail"] + "<br>Hello World! v1 <br> Server's time:<b>" + time.strftime ("%m-%d-%y %I:%M:%S %p")+"</b>"

if __name__ == '__main__':
  app.run()
```

When I navigate to site, I get credential prompt. When credentials are provided, internal server error is reported. From the docker logs:

```
2018-08-01T06:07:42.731887152Z Traceback (most recent call last):
2018-08-01T06:07:42.731891152Z   File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 2292, in wsgi_app
2018-08-01T06:07:42.731934053Z     response = self.full_dispatch_request()
2018-08-01T06:07:42.731937753Z   File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1815, in full_dispatch_request
2018-08-01T06:07:42.731941353Z     rv = self.handle_user_exception(e)
2018-08-01T06:07:42.731944653Z   File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1718, in handle_user_exception
2018-08-01T06:07:42.731948153Z     reraise(exc_type, exc_value, tb)
2018-08-01T06:07:42.731951653Z   File "/usr/local/lib/python3.6/site-packages/flask/_compat.py", line 35, in reraise
2018-08-01T06:07:42.731955053Z     raise value
2018-08-01T06:07:42.731958353Z   File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1813, in full_dispatch_request
2018-08-01T06:07:42.731961653Z     rv = self.dispatch_request()
2018-08-01T06:07:42.731964853Z   File "/usr/local/lib/python3.6/site-packages/flask/app.py", line 1799, in dispatch_request
2018-08-01T06:07:42.731968153Z     return self.view_functions[rule.endpoint](**req.view_args)
2018-08-01T06:07:42.731971453Z   File "/usr/local/lib/python3.6/site-packages/flask_dance/consumer/oauth2.py", line 246, in authorized
2018-08-01T06:07:42.731974853Z     **self.token_url_params
2018-08-01T06:07:42.731978553Z   File "/usr/local/lib/python3.6/site-packages/requests_oauthlib/oauth2_session.py", line 244, in fetch_token
2018-08-01T06:07:42.731981953Z     self._client.parse_request_body_response(r.text, scope=self.scope)
2018-08-01T06:07:42.731985153Z   File "/usr/local/lib/python3.6/site-packages/oauthlib/oauth2/rfc6749/clients/base.py", line 411, in parse_request_body_response
2018-08-01T06:07:42.731988653Z     self.token = parse_token_response(body, scope=scope)
2018-08-01T06:07:42.731991753Z   File "/usr/local/lib/python3.6/site-packages/oauthlib/oauth2/rfc6749/parameters.py", line 379, in parse_token_response
2018-08-01T06:07:42.731995153Z     validate_token_parameters(params)
2018-08-01T06:07:42.732004253Z   File "/usr/local/lib/python3.6/site-packages/oauthlib/oauth2/rfc6749/parameters.py", line 409, in validate_token_parameters
2018-08-01T06:07:42.732007953Z     raise w
2018-08-01T06:07:42.732011153Z Warning: Scope has changed from "user.read" to "User.Read".
```

**Comment by singingwolfboy** [comment]
Looks like Azure has changed the capitalization of the default scope. You should be able to fix this by specifying your scopes explicitly:

```python
blueprint = make_azure_blueprint(
    client_id="foo",
    client_secret="bar",
    scopes=["User.Read"]
)
```

@steven-martins, you're the maintainer of the Azure configuration. [Looks like this line needs to be changed.](https://github.com/singingwolfboy/flask-dance/blob/master/flask_dance/contrib/azure.py#L48). Can you verify this, and make a pull request to do so if I'm correct?

**Comment by leonbrag** [comment]
other then minor correction, it worked,  thanks! it should be scope:

`scope=["User.Read"]`

**Comment by atwalsh** [comment]
I had to change my scope to: 

`User.Read profile email openid`

The warning I received was:

```shell
Warning: Scope has changed from "User.Read" to "User.Read profile email openid".
```

**Comment by jacopofar** [comment]
I get the same error as well, it seems something changed recently because the code previously worked :/

I'm using the `/common/` tenant, but with a company account.

**Comment by yn-coder** [comment]
Hello! May it's similar issue, previous recomendations is not resolve my trouble

```python
blueprint = make_azure_blueprint(
    client_id="---",
    client_secret="+++",
    #scope=["profile email User.Read openid"]
    scope=["User.Read"]
    
)
app.register_blueprint(blueprint, url_prefix="/login")
```

an got this fail on heroku

```
2018-08-06T11:26:06.867566+00:00 app[web.1]: 188.170.82.247 - - [06/Aug/2018:11:26:06 +0000] "GET /t/ HTTP/1.1" 200 8 "-" "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/68.0.3440.84 Safari/537.36"

2018-08-06T11:26:13.382077+00:00 app[web.1]: 188.170.82.247 - - [06/Aug/2018:11:26:13 +0000] "GET / HTTP/1.1" 302 231 "-" "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/68.0.3440.84 Safari/537.36"

2018-08-06T11:26:13.963216+00:00 app[web.1]: 188.170.82.247 - - [06/Aug/2018:11:26:13 +0000] "GET /login/azure HTTP/1.1" 302 811 "-" "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/68.0.3440.84 Safari/537.36"

2018-08-06T11:27:17.298566+00:00 app[web.1]: [2018-08-06 11:27:17,295] ERROR in app: Exception on /login/azure/authorized [GET]

2018-08-06T11:27:17.298583+00:00 app[web.1]: Traceback (most recent call last):

2018-08-06T11:27:17.298585+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/flask/app.py", line 2292, in wsgi_app

2018-08-06T11:27:17.298587+00:00 app[web.1]:     response = self.full_dispatch_request()

2018-08-06T11:27:17.298590+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/flask/app.py", line 1815, in full_dispatch_request

2018-08-06T11:27:17.298592+00:00 app[web.1]:     rv = self.handle_user_exception(e)

2018-08-06T11:27:17.298594+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/flask/app.py", line 1718, in handle_user_exception

2018-08-06T11:27:17.298596+00:00 app[web.1]:     reraise(exc_type, exc_value, tb)

2018-08-06T11:27:17.298598+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/flask/_compat.py", line 35, in reraise

2018-08-06T11:27:17.298601+00:00 app[web.1]:     raise value

2018-08-06T11:27:17.298602+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/flask/app.py", line 1813, in full_dispatch_request

2018-08-06T11:27:17.298604+00:00 app[web.1]:     rv = self.dispatch_request()

2018-08-06T11:27:17.298606+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/flask/app.py", line 1799, in dispatch_request

2018-08-06T11:27:17.298608+00:00 app[web.1]:     return self.view_functions[rule.endpoint](**req.view_args)

2018-08-06T11:27:17.298610+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/flask_dance/consumer/oauth2.py", line 246, in authorized

2018-08-06T11:27:17.298612+00:00 app[web.1]:     **self.token_url_params

2018-08-06T11:27:17.298614+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/requests_oauthlib/oauth2_session.py", line 244, in fetch_token

2018-08-06T11:27:17.298620+00:00 app[web.1]:     self._client.parse_request_body_response(r.text, scope=self.scope)

2018-08-06T11:27:17.298623+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/oauthlib/oauth2/rfc6749/clients/base.py", line 411, in parse_request_body_response

2018-08-06T11:27:17.298624+00:00 app[web.1]:     self.token = parse_token_response(body, scope=scope)

2018-08-06T11:27:17.298626+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/oauthlib/oauth2/rfc6749/parameters.py", line 379, in parse_token_response

2018-08-06T11:27:17.298627+00:00 app[web.1]:     validate_token_parameters(params)

2018-08-06T11:27:17.298629+00:00 app[web.1]:   File "/app/.heroku/python/lib/python3.6/site-packages/oauthlib/oauth2/rfc6749/parameters.py", line 409, in validate_token_parameters

2018-08-06T11:27:17.298630+00:00 app[web.1]:     raise w

2018-08-06T11:27:17.298735+00:00 app[web.1]: Warning: Scope has changed from "User.Read profile openid email" to "User.Read profile openid email".

2018-08-06T11:27:17.299993+00:00 app[web.1]: 188.170.82.247 - - [06/Aug/2018:11:27:17 +0000] "GET /login/azure/authorized?code=M1a450c4a-cf19-7ac2-d41d-6712a1148eb2&state=J78X4zmvnWuiPpN0oa8n1Crn63eDlG HTTP/1.1" 500 291 "https://login.microsoftonline.com/common/oauth2/v2.0/authorize?response_type=code&client_id=e9bf7a9a-13bc-4db8-af9d-f6626ec9705f&redirect_uri=https%3A%2F%2Fcodingteacher.herokuapp.com%2Flogin%2Fazure%2Fauthorized&scope=User.Read+profile+openid+email&state=J78X4zmvnWuiPpN0oa8n1Crn63eDlG" "Mozilla/5.0 (Windows NT 6.1; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/68.0.3440.84 Safari/537.36"
```

**Comment by steven-martins** [comment]
Hello,

I will do some research about these problems of scope this week.

**Comment by yn-coder** [comment]
Here is a working sample for my situation

```python
from flask import Flask, redirect, url_for
from werkzeug.contrib.fixers import ProxyFix
from flask_dance.contrib.azure import make_azure_blueprint, azure

app = Flask(__name__)

app.wsgi_app = ProxyFix(app.wsgi_app)
app.secret_key = "supersekrit"
blueprint = make_azure_blueprint(
    client_id="--",
    client_secret="+++",

    # here is my changes
    scope=["User.Read","profile","openid","email"]  
    
)
app.register_blueprint(blueprint, url_prefix="/login")

@app.route("/")
def index():
    if not azure.authorized:
        return redirect(url_for("azure.login"))
    
    try:
        resp = azure.get("/v1.0/me")
        assert resp.ok
        print(resp.json())
        # here is my changes
        return "You are {mail} on Azure AD".format(mail=resp.json()["userPrincipalName"])

```

2 changes from tour sample

1. Scope is a comma-delimited list `scope=["User.Read","profile","openid","email"]`
2. graph response is json ` {'@odata.context': 'https://graph.microsoft.com/v1.0/$metadata#users/$entity', 'displayName': 'Y N', 'surname': 'N', 'givenName': 'Y', 'id': 'ac0000000000', 'userPrincipalName': 'mail_box@outlook.com', 'businessPhones': [], 'jobTitle': None, 'mail': None, 'mobilePhone': None, 'officeLocation': None, 'preferredLanguage': None}`

**Comment by steven-martins** [comment]
Thank you @yn-coder .

I made some tests and "User.Read" can be enough for personal accounts.
But your modifications (`scope=["User.Read","profile","openid","email"]`) work for the both cases (personal or entreprise account). So I updated (#155) the code consequently.

**Comment by daenney** [comment]
I believe this is fixed with #155 getting merged, so I'm closing the issue. If that's not the case please re-open it.
